ejb3/src/main/java/org/jboss/as/ejb3/timerservice/TimerServiceImpl.java

```diff
@@ -1181,6 +1212,55 @@ public void afterCompletion(final int status) {
         }
     }
 
+    private class TaskPostPersist extends java.util.TimerTask {
+        private final TimerImpl timer;
+        private long delta = 0;
+
+        TaskPostPersist(TimerImpl timer) {
+            this.timer = timer;
+        }
+
+        TaskPostPersist(TimerImpl timer, long delta) {
+            this.timer = timer;
+            this.delta = delta;
+        }
+
+        @Override
+        public void run() {
+            final ExecutorService executor = executorServiceInjectedValue.getOptionalValue();
+            if (executor != null) {
+                executor.submit(this::persistTimer);
+            }
+        }
+
+        void persistTimer() {
+            final ContextTransactionManager transactionManager = ContextTransactionManager.getInstance();
+            try {
+                transactionManager.begin();
+                timerPersistence.getValue().persistTimer(timer);
+                transactionManager.commit();
+            } catch (Exception e) {
+                try {
+                    transactionManager.rollback();
+                } catch (Exception ee) {
+                    // omit;
+                }
+                EJB3_TIMER_LOGGER.exceptionRunningTimerTask(timer, timer.getTimedObjectId(), e);
+                Date nextExpiration = timer.getNextExpiration();
+                long nextExpirationDelay;
+                if (nextExpiration != null &&
+                        (nextExpirationDelay = nextExpiration.getTime() - System.currentTimeMillis()) > delta) {
+                    if (delta == 0L) {
+                        delta = nextExpirationDelay / (1L + MAX_RETRY.longValue());
+                    }
+                    timerInjectedValue.getValue().schedule(new TaskPostPersist(timer, delta), delta);
```

Comment:
It does not look like you are actually tracking the number of retries? MAX_RETRIES seems to only be used to calculate the delta?

Reply:
Absolutely! The issue is that the persistence could be fired again on the next timeout or in the repeat stage (in case of exceptions during timer invocation). So, we do not want to have several concurrent processes for storing current state. 

I have made changes to prevent repetitions after the next (pristine) timer expiration. So, any repetitions (not more than MAX_RETRY) should be finished before any other processes, which could change expiration and invoke persist method.

The only state, which does not change expiration is RETRY_TIMEOUT. I omit repetitions for this state.

Please, see my changes in the next commit.